parl/remote/decorator.py

```diff
@@ -0,0 +1,189 @@
+
+@virtual 
+Class Simulator(object):
+    ...
+
+sim = Simulator()
+sim.remote_run(server_ip='172.18.202.45', port=8001)
+
+"""
+
+
+def virtual(cls, location='client'):
+    """
+    Class wrapper for wrapping a normal class as a remote class that can run in different machines.
+    Two kinds of wrapper are provided for the client as well as the server.
+
+    Args:
+        location(str): specify which wrapper to use, available locations: client/server.
+                       users are expected to use `client`.
+    """
+    assert location in ['client', 'server'], \
+        'Remote Class has to be placed at client side or server side.'
+
+    class ClientWrapper(object):
+        """
+        Wrapper for remote class at client side. After the decoration, 
+        the initial class is able to be called to run any function at sever side.
+        """
+
+        def __init__(self, *args):
+            """
+            Args:
+                args: arguments for the initialisation of the initial class.
+            """
+            self.unwrapped = cls(*args)
+            self.conect_socket = None
+            self.reply_socket = None
+
+        def create_reply_socket(self):
+            """
+            In fact, we have also a socket server in client side. This server keeps running 
+            and waits for requests (e.g. call a function) from server side.
+            """
+            client_ip = get_ip_address()
+            context = zmq.Context()
+            socket = context.socket(zmq.REP)
+            free_port = None
+            for port in range(6000, 8000):
+                try:
+                    socket.bind("tcp://*:{}".format(port))
+                    logger.info(
+                        "[create_reply_socket] free_port:{}".format(port))
+                    free_port = port
+                    break
+                except zmq.error.ZMQError:
+                    logger.warn(
+                        "[create_reply_socket]cannot bind port:{}, retry".
+                        format(port))
+            if free_port is not None:
+                return socket, client_ip, free_port
+            else:
+                logger.error(
+                    "cannot find any available port from 6000 to 8000")
+                sys.exit(1)
+
+        def connect_server(self, server_ip, server_port):
+            """
+            create the connection between client side and server side.
+
+            Args:
+                server_ip(str): the ip of the server.
+                server_port(int): the connection port of the server.
+            """
+            self.reply_socket, local_ip, local_port = self.create_reply_socket(
+            )
+
+            logger.info("connecting {}:{}".format(server_ip, server_port))
+            context = zmq.Context()
+            socket = context.socket(zmq.REQ)
+            socket.connect("tcp://{}:{}".format(server_ip, server_port))
+            client_id = np.random.randint(int(1e18))
```

Comment:
client_id is not necessary?

Reply:
Not currently. As we have discussed, the approach we use now requires to find an available port for client. However, this approach is not suitable to run in the hadoop cluster.
Another approach is building the connection through the method that not require a port. In this approach, a client ID is necessary.